worksheet.rb

```diff
@@ -3,32 +3,157 @@
+        date: "4th Feb 2016",
+        cost: 10,
+        rider_id: "RD0022",
+        rating: 4
+    },
+
+    {
+        driver_id: "DR0002",
+        date: "4th Feb 2016",
+        cost: 15,
+        rider_id: "RD0013",
+        rating: 1
+    },
 
-# Setup the entire data structure:
-# based off of the notes you have above, create the
-# and manually write in data presented in rides.csv
-# You should be copying and pasting the literal data
-# into this data structure, such as "DR0004"
-# and "3rd Feb 2016" and "RD0022"
+    {
+        driver_id: "DR0003",
+        date: "5th Feb 2016",
+        cost: 50,
+        rider_id: "RD0003",
+        rating: 2
+    },
+
+    {
+        driver_id: "DR0002",
+        date: "5th Feb 2016",
+        cost: 35,
+        rider_id: "RD0066",
+        rating: 3
+    },
+
+    {
+        driver_id: "DR0004",
+        date: "5th Feb 2016",
+        cost: 20,
+        rider_id: "RD0073",
+        rating: 5
+    },
+
+    {
+        driver_id: "DR0001",
+        date: "5th Feb 2016",
+        cost: 45,
+        rider_id: "RD0003",
+        rating: 2
+    }
+]
 
 ########################################################
 # Step 4: Total Driver's Earnings and Number of Rides
-
 # Use an iteration blocks to print the following answers:
+
+def driver_info(rides, driver)
+  # creating the array of hashes with all rides for each driver
+  driver_rides = []
+  rides.each do |ride|
+    if ride[:driver_id] == driver
+      driver_rides << ride
+    end
+  end
+
+  #calculating the summary of costs of all rides for each driver
+  driver_earned = 0
+  driver_rides.each do |ride|
+    driver_earned += ride[:cost]
+  end
+
+  #calcualting the average rating for each driver, reusing driver_rides array
+  driver_rating_summary = 0
+  driver_rides.each do |ride|
+    driver_rating_summary += ride[:rating]
+  end
+  average_rating = driver_rating_summary.to_f / driver_rides.length
+  return {driver_id: driver, earned: driver_earned, average_rating: average_rating, rides: driver_rides.length}
+end
+
+#iterates through the array driver_infos and calling the method for each driver id with calculations
+all_drivers_ids = ["DR0001", "DR0002", "DR0003", "DR0004"]
+driver_infos = all_drivers_ids.map {|id| driver_info(rides, id)}
+
 # - the number of rides each driver has given
+driver_infos.each {|info| puts "The driver #{info[:driver_id]} has given: #{info[:rides]} rides."}
+
 # - the total amount of money each driver has made
+driver_infos.each {|info| puts "The total amount of money driver #{info[:driver_id]} has made: #{info[:earned]}"}
+
 # - the average rating for each driver
+driver_infos.each {|info| puts "The average rating for driver #{info[:driver_id]} is: #{info[:average_rating]}"}
+
 # - Which driver made the most money?
-# - Which driver has the highest average rating?
\ No newline at end of file
+driver_id_earned_max = driver_infos.max { |driver1, driver2| driver1[:earned] <=> driver2[:earned] }
```

Comment:
This would be a little cleaner using `max_by`:

```suggestion
driver_id_earned_max = driver_infos.max_by { |driver| driver[:earned] }
```